src/expr.rs

```diff
@@ -320,6 +331,61 @@ fn rewrite_if_else(context: &RewriteContext,
     Some(result)
 }
 
+fn single_line_if_else(context: &RewriteContext,
+                       pat_expr_str: &str,
+                       if_node: &ast::Block,
+                       else_block_opt: Option<&ast::Expr>,
+                       width: usize)
+                       -> Option<String> {
+    let else_block = try_opt!(else_block_opt);
+    let fixed_cost = "if  {  } else {  }".len();
+
+    if let ast::ExprBlock(ref else_node) = else_block.node {
+        if !is_simple_block(if_node, context.codemap) ||
+           !is_simple_block(else_node, context.codemap) || pat_expr_str.contains('\n') {
+            return None;
+        }
+
+        let new_width = try_opt!(width.checked_sub(pat_expr_str.len() + fixed_cost));
+        let if_expr = if_node.expr.as_ref().unwrap();
+        let if_str = try_opt!(if_expr.rewrite(context, new_width, 0));
+
+        let new_width = try_opt!(new_width.checked_sub(if_str.len()));
+        let else_expr = else_node.expr.as_ref().unwrap();
+        let else_str = try_opt!(else_expr.rewrite(context, new_width, 0));
+
+        // FIXME: this check shouldn't be necessary. Rewrites should either fail
+        // or wrap to a newline when the object does not fit the width.
+        let fits_line = fixed_cost + pat_expr_str.len() + if_str.len() + else_str.len() <= width;
+
+        if fits_line && !if_str.contains('\n') && !else_str.contains('\n') {
+            return Some(format!("if {} {{ {} }} else {{ {} }}", pat_expr_str, if_str, else_str));
+        }
+    }
+
+    None
+}
+
+// Checks that a block contains no statements, an expression and no comments.
+fn is_simple_block(block: &ast::Block, codemap: &CodeMap) -> bool {
+    if !block.stmts.is_empty() || block.expr.is_none() {
+        return false;
+    }
+
+    let expr = block.expr.as_ref().unwrap();
+    let pre_expr_span = mk_sp(block.span.lo + BytePos(1), expr.span.lo);
+    let pre_expr_snippet = codemap.span_to_snippet(pre_expr_span).unwrap();
+
+    if pre_expr_snippet.trim().len() > 0 {
+        return false;
+    }
+
+    let post_expr_span = mk_sp(expr.span.hi + BytePos(1), block.span.hi - BytePos(1));
+    let post_expr_snippet = codemap.span_to_snippet(post_expr_span).unwrap();
+
+    post_expr_snippet.trim().len() == 0
```

Comment:
Could you check for `//` or `/*` in the span instead of checking lengths? Checking lengths feels a bit fragile.